changefeedccl: disallow changfeeds on regional by row tables

stevendanna opened the pull request:
To fully support changefeeds, we need to make a number of technical-
and product-level decisions and changes. Until those changes are made,
we want to disable changefeeds on regional by row tables.

This change ads a new check in the table validator that will prevent
new changefeeds from being started on a table where
IsLocalityRegionalByRow() is true.

Further, existing feeds will fails with a non-retriable error.

Fixes #63239

Release note (enterprise change): Changefeeds will now fail on any
regional by row table with an error such as:

    CHANGEFEED cannot target REGIONAL BY ROW tables: TABLE_NAME

This is to prevent unexpected changefeed behavior until we can offer
full support for regional by row tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63306)
<!-- Reviewable:end -->

stevendanna commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -214,7 +214,14 @@ func (f *kvFeed) run(ctx context.Context) (err error) {
+			// error. The schemafeed will fail with an
+			// non-retriable error, meaning we likely
+			// return right after runUntilTableEvent
+			// above.
```
Note: because of this, if we want the absolutely most minimal change, we can avoid any changes to the event_table_filter and this code and just rely on the error coming out of validate.  I realized that the validate changes were  enough after writing this code and then thought it might be good to keep it in case the validate code changes, but given that we have a test for in-flight changefeeds failing, perhaps we can drop it.

stevendanna replied:
I've split this into a second commit to make it easier to drop if we want to.

stevendanna commented:
@amruss I used a slightly different error message than what you recommended to be consistent with the error messages we output for other unsupported tables such as those with column families.  Let me know if you'd prefer I change it to your recommended language.

amruss commented:
>  I used a slightly different error message than what you recommended to be consistent with the error messages we output for other unsupported tables such as those with column families. Let me know if you'd prefer I change it to your recommended language.

LGTM - probably better to be consistent here

stevendanna committed:
changefeedccl: disallow changfeeds on regional by row tables

To fully support changefeeds, we need to make a number of technical-
and product-level decisions and changes. Until those changes are made,
we want to disable changefeeds on regional by row tables.

This change ads a new check in the table validator that will prevent
new changefeeds from being started on a table where
IsLocalityRegionalByRow() is true.

Further, existing feeds will fails with a non-retriable error.

Fixes #63239

Release note (enterprise change): Changefeeds will now fail on any
regional by row table with an error such as:

    CHANGEFEED cannot target REGIONAL BY ROW tables: TABLE_NAME

This is to prevent unexpected changefeed behavior until we can offer
full support for regional by row tables.

stevendanna committed:
changefeed: detect regional by row changes in kv_feed

This isn't strictly necessary as the schemafeed currently fails with
an unretriable error if it fails to validate the table descriptor and
regional by row tables will fail to validate.

However, it might be useful in general to be able to detect
unsupported schema migrations at this location since other schema
changes that we might want to disallow will not necessarily force a
table descriptor validation failure.

Release note: None

stevendanna commented:
The first commit is all that is strictly needed, so it might be good to go commit-by-commit.

miretskiy commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -1172,6 +1177,59 @@ func TestChangefeedAuthorization(t *testing.T) {
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRace(t)
+	skip.UnderShort(t)
```
does the test take long time to run?

stevendanna replied:
Not the fastest but not the slowest either:

```
--- PASS: TestChangefeedFailOnRBRChange (1.32s)
    --- PASS: TestChangefeedFailOnRBRChange/sinkless (0.48s)
        --- PASS: TestChangefeedFailOnRBRChange/sinkless/regional_by_row (0.28s)
    --- PASS: TestChangefeedFailOnRBRChange/enterprise (0.82s)
        --- PASS: TestChangefeedFailOnRBRChange/enterprise/regional_by_row (0.53s)
```

miretskiy commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -1172,6 +1177,59 @@ func TestChangefeedAuthorization(t *testing.T) {
+		sqlDB := sqlutils.MakeSQLRunner(db)
+		sqlDB.Exec(t, "SET CLUSTER SETTING changefeed.experimental_poll_interval = '200ms'")
+		sqlDB.Exec(t, "SET CLUSTER SETTING kv.closed_timestamp.target_duration = '50ms'")
+		sqlDB.Exec(t, "SET CLUSTER SETTING kv.closed_timestamp.close_fraction = .99")
```
I don't think these 3 are needed, are they?  I thought sinkless+enterprise test below do that already.

stevendanna replied:
I removed 2 of the 3.  The lower target_duration does seem to speed up this test substantially and right now we have a default of 1s.

stevendanna commented:
bors r=miretskiy
